be/src/exec/pipeline/scan/connector_scan_operator.cpp

```diff
@@ -502,27 +555,73 @@ ConnectorScanOperatorIOTasksMemLimiter* ConnectorChunkSource::_get_io_tasks_mem_
 void ConnectorChunkSource::close(RuntimeState* state) {
     if (_closed) return;
 
+    MemTracker* mem_tracker = state->query_ctx()->connector_scan_mem_tracker();
+    mem_tracker->release(_request_mem_tracker_bytes);
+
     ConnectorScanOperatorIOTasksMemLimiter* limiter = _get_io_tasks_mem_limiter();
     limiter->update_running_chunk_source_count(-1);
-    limiter->update_chunk_source_mem_bytes(_data_source->estimated_mem_usage());
+    uint64_t chunk_num = std::min<uint64_t>(state->chunk_size(), _chunk_rows_read);
+    int64_t chunk_source_mem_bytes = _data_source->estimated_mem_usage() + avg_chunk_mem_bytes() * chunk_num;
+    limiter->update_chunk_source_mem_bytes(chunk_source_mem_bytes);
+
+    [[maybe_unused]] auto build_debug_string = [&]() {
+        std::stringstream ss;
+        ss << "try_mem_tracker. query_id = " << print_id(state->query_id())
+           << ", op_id = " << _scan_op->get_plan_node_id() << "/" << _scan_op->get_driver_sequence()
+           << ", release. this = " << (void*)this << ", request mem bytes = " << _request_mem_tracker_bytes
+           << ", chunk source mem bytes = " << chunk_source_mem_bytes << ", chunk mem bytes = " << avg_chunk_mem_bytes()
+           << " * " << chunk_num << ", data source mem usage = " << _data_source->estimated_mem_usage();
+        return ss.str();
+    };
+    VLOG_OPERATOR << build_debug_string();
 
     _closed = true;
     _data_source->close(state);
 }
 
-Status ConnectorChunkSource::_open_data_source(RuntimeState* state) {
+Status ConnectorChunkSource::_open_data_source(RuntimeState* state, bool* mem_alloc_failed) {
     if (_opened) {
         return Status::OK();
     }
 
+    ConnectorScanOperatorIOTasksMemLimiter* limiter = _get_io_tasks_mem_limiter();
+    _request_mem_tracker_bytes = limiter->chunk_source_mem_bytes.load();
+
+    [[maybe_unused]] auto build_debug_string = [&](const std::string action) {
+        std::stringstream ss;
+        ss << "try_mem_tracker. query_id = " << print_id(state->query_id())
+           << ", op_id = " << _scan_op->get_plan_node_id() << "/" << _scan_op->get_driver_sequence() << ", " << action
+           << ". this = " << (void*)this << ", value = " << _request_mem_tracker_bytes;
+        return ss.str();
+    };
+
+    MemTracker* mem_tracker = state->query_ctx()->connector_scan_mem_tracker();
+
+    int retry = 3;
+    while (retry > 0) {
+        retry--;
+        *mem_alloc_failed = false;
+        if (mem_tracker->try_consume(_request_mem_tracker_bytes) == nullptr) {
+            break;
+        }
+        *mem_alloc_failed = true;
+        // VLOG_OPERATOR << build_debug_string("alloc failed");
+        std::this_thread::sleep_for(std::chrono::milliseconds(10));
```

Comment:
Here we retry 3 times, and sleep 10ms if mem alloc failed.

Reply:
I did not think throught about pros and cons of this strategy.  

From intutition, frequent submission of chunk source because of mem alloc failed is not good idea.